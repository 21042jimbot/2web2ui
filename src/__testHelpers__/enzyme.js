export const findButton = (wrapper, buttonLabel) =>
  wrapper.findWhere(
    (node) => node.type() === 'button' && new RegExp(buttonLabel).test(node.text())
  );
