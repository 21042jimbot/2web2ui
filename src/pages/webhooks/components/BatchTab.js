import React, { Component } from 'react';
import { connect } from 'react-redux';
import _ from 'lodash';
import { getBatches } from 'src/actions/webhooks';
import { PanelSectionTableCollection } from 'src/components/collection';
import { PanelSectionLoading } from 'src/components/loading';
import { Button, Panel } from 'src/components/matchbox';
import { Empty } from 'src/components';
import { selectWebhookBatches } from 'src/selectors/webhooks';

const columns = [
  { label: 'Delivery Time', sortKey: 'ts' }, //This is timestamp
  { label: 'Batch ID', sortKey: 'batch_id' },
  { label: 'Status', sortKey: 'status' },
  { label: 'Attempt #', sortKey: 'attempts' },
  { label: 'Response', sortKey: 'response_code' },
];

const getRowData = batch => [
  batch.formatted_time,
  batch.batch_id,
  batch.status,
  batch.attempts,
  batch.response_code,
];

export class BatchTab extends Component {
  componentDidMount() {
    this.refreshBatches();
  }

  refreshBatches = () => {
    const { webhook, getBatches } = this.props;
    const { id, subaccount } = webhook;

    return getBatches({ id, subaccount });
  };

  renderBatches() {
    const { batches, batchesLoading } = this.props;

    if (batchesLoading) {
      return <PanelSectionLoading />;
    }

    if (_.isEmpty(batches)) {
      return <Empty message="There are no batches for your webhook" />;
    }

    return (
      <PanelSectionTableCollection
        columns={columns}
        rows={batches}
        getRowData={getRowData}
        pagination={true}
        defaultSortColumn="ts"
        defaultSortDirection="desc"
      />
    );
  }

  render() {
    const { batchesLoading } = this.props;
    const buttonText = batchesLoading ? 'Refreshing...' : 'Refresh Batches';

    return (
      <>
        <Panel.Section>
          <Button variant="primary" disabled={batchesLoading} onClick={this.refreshBatches}>
            {buttonText}
          </Button>
        </Panel.Section>
        {this.renderBatches()}
      </>
    );
  }
}

const mapStateToProps = state => ({
  batches: selectWebhookBatches(state),
  batchesLoading: state.webhooks.batchesLoading,
});

export default connect(mapStateToProps, { getBatches })(BatchTab);
