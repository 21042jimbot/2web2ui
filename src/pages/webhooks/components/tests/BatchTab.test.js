import { shallow } from 'enzyme';
import React from 'react';

import { BatchTab } from '../BatchTab';

describe('Webhook Component: Batch Status Tab', () => {
  const props = {
    getBatches: jest.fn(),
    id: 'my-id',
    batches: [
      {
        formatted_time: 'so-formatted',
        batch_id: '243423423423',
        status: 'p',
        attempts: 1,
        response_code: 200
      },
      {
        formatted_time: 'so-formatted-2',
        batch_id: '996969545',
        status: 'f',
        attempts: 4,
        response_code: 500
      }
    ],
    batchesLoading: false
  };

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<BatchTab {...props} />);
  });

  it('should render batch status tab with table data', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should show button text as Refreshing while refreshing data', () => {
    wrapper.setState({ refreshing: true });
    wrapper.setProps({ batchesLoading: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('should show empty message when no batches exist', () => {
    wrapper.setProps({ batches: []});
    expect(wrapper).toMatchSnapshot();
  });

  it('should show panel loading when batches are loading', () => {
    wrapper.setProps({ batchesLoading: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('should call getBatches in mounting', () => {
    const instance = wrapper.instance();
    instance.componentDidMount();
    expect(instance.props.getBatches).toHaveBeenCalledWith('my-id');
  });

  it('should getBatches when refreshing', () => {
    const instance = wrapper.instance();
    const stateSpy = jest.spyOn(instance, 'setState');
    instance.refreshBatches();
    expect(instance.props.getBatches).toHaveBeenCalledWith('my-id');
    expect(stateSpy).toHaveBeenCalledWith({ refreshing: true });
  });

});
