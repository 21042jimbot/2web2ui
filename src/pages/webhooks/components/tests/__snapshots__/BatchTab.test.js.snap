// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webhook Component: Batch Status Tab should render batch status tab with table data 1`] = `
<Panel>
  <Panel.Section>
    <Button
      disabled={false}
      onClick={[Function]}
      primary={true}
      size="small"
    >
      Refresh Batches
    </Button>
  </Panel.Section>
  <TableCollection
    columns={
      Array [
        Object {
          "label": "Delivery Time",
          "sortKey": "ts",
        },
        Object {
          "label": "Batch ID",
          "sortKey": "batch_id",
        },
        Object {
          "label": "Status",
          "sortKey": "status",
        },
        Object {
          "label": "Attempt #",
          "sortKey": "attempts",
        },
        Object {
          "label": "Response",
          "sortKey": "response_code",
        },
      ]
    }
    defaultSortColumn="ts"
    defaultSortDirection="desc"
    getRowData={[Function]}
    pagination={true}
    rows={
      Array [
        Object {
          "attempts": 1,
          "batch_id": "243423423423",
          "formatted_time": "so-formatted",
          "response_code": 200,
          "status": "p",
          "ts": 1,
        },
        Object {
          "attempts": 4,
          "batch_id": "996969545",
          "formatted_time": "so-formatted-2",
          "response_code": 500,
          "status": "f",
          "ts": 2,
        },
      ]
    }
  />
</Panel>
`;

exports[`Webhook Component: Batch Status Tab should show button text as Refreshing while refreshing data 1`] = `
<Panel>
  <Panel.Section>
    <Button
      disabled={true}
      onClick={[Function]}
      primary={true}
      size="small"
    >
      Refreshing...
    </Button>
  </Panel.Section>
  <PanelLoading
    minHeight="400px"
  />
</Panel>
`;

exports[`Webhook Component: Batch Status Tab should show empty message when no batches exist 1`] = `
<Panel>
  <Panel.Section>
    <Button
      disabled={false}
      onClick={[Function]}
      primary={true}
      size="small"
    >
      Refresh Batches
    </Button>
  </Panel.Section>
  <Empty
    message="There are no batches for your webhook"
  />
</Panel>
`;
